langchain/chat_models/vertexai.py

```diff
@@ -0,0 +1,141 @@
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import root_validator
+
+from langchain.callbacks.manager import (
+    AsyncCallbackManagerForLLMRun,
+    CallbackManagerForLLMRun,
+)
+from langchain.chat_models.base import BaseChatModel
+from langchain.llms.vertexai import _VertexAICommon
+from langchain.schema import (
+    AIMessage,
+    BaseMessage,
+    ChatGeneration,
+    ChatResult,
+    HumanMessage,
+    SystemMessage,
+)
+from langchain.utilities.vertexai import raise_vertex_import_error
+
+
+@dataclass
+class _MessagePair:
+    """InputOutputTextPair represents a pair of input and output texts."""
+
+    question: HumanMessage
+    answer: AIMessage
+
+
+@dataclass
+class _ChatHistory:
+    """InputOutputTextPair represents a pair of input and output texts."""
+
+    history: List[_MessagePair] = field(default_factory=list)
+    system_message: Optional[SystemMessage] = None
+
+
+def _parse_chat_history(history: List[BaseMessage]) -> _ChatHistory:
+    """Parses a sequence of messages into history.
+
+    A sequency should be either (SystemMessage, HumanMessage, AIMessage,
+    HumanMessage, AIMessage, ...) or (HumanMessage, AIMessage, HumanMessage,
+    AIMessage, ...).
+    """
+    if not history:
+        return _ChatHistory()
+    first_message = history[0]
+    system_message = first_message if isinstance(first_message, SystemMessage) else None
+    chat_history = _ChatHistory(system_message=system_message)
+    messages_left = history[1:] if system_message else history
+    if len(messages_left) % 2 != 0:
+        raise ValueError(
+            f"Amount of messages in history should be even, got {len(messages_left)}!"
+        )
+    for question, answer in zip(messages_left[::2], messages_left[1::2]):
+        if not isinstance(question, HumanMessage) or not isinstance(answer, AIMessage):
+            raise ValueError(
+                "A human message should follow a bot one, "
+                f"got {question.type}, {answer.type}."
+            )
+        chat_history.history.append(_MessagePair(question=question, answer=answer))
+    return chat_history
+
+
+class ChatVertexAI(_VertexAICommon, BaseChatModel):
+    """Wrapper around Vertex AI large language models.
+
+    To use, you should have the
+    ``google.cloud.aiplatform.private_preview.language_models`` python package
+    installed.
+    """
+
+    model_name: str = "chat-bison"
+    chat: Any = None  #: :meta private:
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validates that the python package exists in environment."""
+        cls._try_init_vertexai(values)
+        try:
+            from vertexai.preview.language_models import ChatModel
+        except ImportError:
+            raise_vertex_import_error()
+        values["client"] = ChatModel.from_pretrained(values["model_name"])
+        return values
+
+    def send_message(
```

Comment:
It seems like publicly exposing `send_message`, `chat`, and `start_chat` can lead to unintuitive usage where history is saved exclusively when using `send_message` and `start_chat` or using `chat` directly, but then cleared when using `generate`. My understanding is that langchain has it's own memory representation, we may want to differ to that in all cases unless there is precedence for these memory side effects elsewhere in the library.

Reply:
Langchain's own memory representation would work perfectly with `generate`, I don't see any conflict here.
Nevertheless, if you feel like Vertex SDK methods shouldn't be exposed through Langchain, I'm fine with it. My line of reasoning was that since the Vertex object keeps an internal history, it would be good to expose it through additional methods. Now I just re-create `_ChatSession` each time.